File: csharp/Subscription.Api/Controllers/SubscriberController.cs
using Microsoft.AspNetCore.Mvc;
using Subscription.Api.Models;
using Subscription.Api.Services;

namespace Subscription.Api.Controllers;

[ApiController]
[Route("api/[controller]")]
public class SubscriberController(IEmailValidationService emailValidationService, SubscriberService subscriberService, ILogger<SubscriberController> logger) : ControllerBase
{
    [HttpPost("register")]
    public async Task<ActionResult<SubscriberResponse>> Register([FromBody] RegisterSubscriberRequest request)
    {
        if (!ModelState.IsValid)
        {
            return BadRequest(ModelState);
        }

        logger.LogInformation("Registering new subscriber with email: {Email}", request.Email);

        // Validate email safety using external service
        var isEmailSafe = await emailValidationService.IsEmailSafeAsync(request.Email);

        if (!isEmailSafe)
        {
            logger.LogWarning("Email validation failed for: {Email}", request.Email);
            return BadRequest("Email is not safe or valid according to validation service");
        }

        var subscriber = await subscriberService.CreateSubscriberAsync(request.Email, request.Name);

        if (subscriber == null)
        {
            logger.LogError("Failed to create subscriber for email: {Email}", request.Email);
            return StatusCode(500, "Failed to create subscriber");
        }

        var response = new SubscriberResponse(subscriber.Id, request.Email, subscriber.CreatedAt);

        logger.LogInformation("Successfully registered subscriber: {Id} with email: {Email}", response.Id, response.Email);

        return CreatedAtAction(nameof(Register), new { id = response.Id }, response);
    }
}

File: typescript/src/subscription/subscriber-controller.ts
import {
  BadRequestException,
  Body,
  Controller,
  Inject,
  Post,
} from '@nestjs/common'
import { EmailValidationService } from './email-validation-service'
import type { ISubscriberService } from './subscriber-service'

@Controller()
export class SubscriberController {
  constructor(
    @Inject(EmailValidationService)
    readonly emailValidationService: EmailValidationService,
    @Inject('ISubscriberService')
    readonly subscriberService: ISubscriberService,
  ) {}

  @Post()
  async register(
    @Body() body: { name: string; email: string; websiteUri: string },
  ) {
    //todo validate body

    if (!(await this.emailValidationService.isEmailSafe(body.email))) {
      throw new BadRequestException('Unsafe email address')
    }

    // call service to create subscriber
    const {
      registeredAt: _registeredAt,
      updatedAt: _updatedAt,
      ...storedSubscriber
    } = await this.subscriberService.createSubscriberAsync(
      body.email,
      body.name,
      body.websiteUri,
    )

    // return http created with the subscriber information
    return storedSubscriber
  }
}
